### ISSUE TITLE
High CPU consumption of Kubelet 1.1

### ISSUE THREAD

**Issue by 2opremio** [kubernetes/kubernetes#19658]
I am running a Kubenetes 1.1.1 cluster in AWS (Docker 1.7.1), formed by [m3.large](https://aws.amazon.com/ec2/instance-types/#M3) machines.

In some of the cluster nodes kubelet is consuming more than 50% of the CPU. This seems unexpectedly high since the number of pods is fairly low (less than 20 per node).

Here's kubelet's profile call graph from one of the affected nodes (obtained with `go tool pprof http://localhost:10248/debug/pprof/profile`):

![profile](https://cloud.githubusercontent.com/assets/2362916/12332236/39b24826-bae5-11e5-989e-444200d59843.png)

And here's the raw profile: http://filebin.ca/2TRxlc18BQaO/pprof.localhost10248.samples.cpu.001.pb.gz

The CPU consumption has grown over time, but so has the number of pods. I am not sure if I am not sure whether I am experiencing https://github.com/kubernetes/kubernetes/issues/15805#issuecomment-148963658 but it seems likely. Can you guys confirm?

**Comment by 2opremio** [comment]
In case it helps, here are the kubelet goroutines (obtained with `curl http://localhost:10248/debug/pprof/goroutine?debug=2`): http://pastebin.com/9pTGNtZ0

**Comment by 2opremio** [comment]
And here is the version info:

```
Server Version: version.Info{Major:"1", Minor:"1", GitVersion:"v1.1.1", GitCommit:"92635e23dfafb2ddc828c8ac6c03c7a7205a84d8", GitTreeState:"clean"}
```

After reading https://github.com/kubernetes/kubernetes/issues/15805#issuecomment-148963658 more carefully, I don't think that's the issue since I am running 1.1.1 and ...

> This does not affect v1.1 because go-dockerclient was only updated recently (#15127)

**Comment by 2opremio** [comment]
Here's a spanshot of kubelet's goroutines, obtained with `curl http://localhost:10248/debug/pprof/goroutine?debug=2`: http://pastebin.com/9pTGNtZ0

**Comment by 2opremio** [comment]
I run a couple more profiles and all consistently show that most of the execution is spent in cadvisor's  [scanInterfaceStats](https://github.com/kubernetes/kubernetes/blob/v1.1.1/Godeps/_workspace/src/github.com/google/cadvisor/container/libcontainer/helpers.go#L150)

**Comment by 2opremio** [comment]
OK, it seems that the these performance issues should had been fixed in master:

- 869cc9fbdab560a1c5eaaf4793020053f802cfbd removes the regexp matching
- 041ab17a677573ef0d67d024860a4e69f1faf5be removes the Sscanf invocation

I am not familiar with the kubernetes release process. Those two commits don't have any associated tag yet, can I expect them to be released with kubernetes 1.2?

**Comment by yujuhong** [comment]
> OK, it seems that the these performance issues should had been fixed in master:
> 
> 869cc9f removes the regexp matching (part of v1.2.0-alpha.5)
> 041ab17 removes the Sscanf invocation (part of v1.2.0-alpha.6)
> I am not familiar with the kubernetes release process but I assume they will be released with kubernetes 1.2

@2opremio, I believe most of the cadvisor fixes have been cherry-picked to the 1.1 branch (#18182). You can consider upgrading to the latest release (1.1.4) to see if the problem has been fixed. Thanks!

**Comment by vishh** [comment]
Thanks for the detailed issue! The cAdvisor bugs you mentioned will be
included as part of v1.2, in addition to other fixes.

**Comment by 2opremio** [comment]
OK, I will close and reopen if the problem persists after upgrading Kubernetes.
